# 3803ICT Big Data Analysis Assignment

Joshua Russell | s5057545 | joshua.russell2@griffithuni.edu.au 

Joshua Mitchell | s5055278 | joshua.mitchell4@griffithuni.edu.au


## Part 1 - Data Preparation and Preprocessing

## Introduction

Data preparation and preprocessing involves three main processes; data storage, data normailzation and data cleaning. We will therefore introduce and discuss these concepts here.

#### Data Storage
Data storage in concerned with the way in which data is stored. Within this investigation, the Python library Pandas will be the primary method used for storing, processing and manipulating data. 

#### Data Normailzation
Within the context of this course, data normailzation encapsulates data scaling. Consequently, scaling will be discussed as a data normailzation technique. However, there is two important distingushing factors that should be outlined. Data scaling is the process by which data is *scaled* into a specific range, such as between 0 and 1. The meaning behind this transformation is to value numerical features with the same importance when using models that implement distance or error functions. To provide more intuition behind this matter, we will discuss an example. 

Say we were analysing the salaries of American university professors with those of Korean professors. For simplicity, we will state that the exchange between USD and Korean won is 1 to 1000. If we were to use an analysis method such as KNN with euclidean distance as the distance function, 1 USD would be weighted the same as 1 Korean won. Our distance function would consequently be misrepresenting the relationship between the two currencies, giving a *bias* to one of the variables. We therefore scale our data such that variables can be properly compared with one another.

Data scaling involves the transformation of the scale of variables into a specific range. The process on data normalization is concerned with a much broader view, looking at how samples are distributed among the entire dataset. Many statistical analysis techniques assume that the data is normally distributed. Datasets that contain skewed or other non-normal distributions of numerical data will need to be transformed into normal distributions such that these techniques can provide meaningful insights about the data. This process of transforming data into a distribution that can be regarded as being normal is data normalization. 

#### Data Cleaning
The process of data cleaning involves finding data with incorrect formatting, data with missing values, outliers, data that is erroneous, inconsistent, irrelevant and malicious, and either correcting or removing those samples. The types of data mentioned here are collectively known as "dirty" data. We will briefly describe the types of dirty data below.

- **Formatting**: data containing the same information with inconsistent formatting
- **Missing data**: data samples with missing variable information
- **Erroneous data**: data that is erroneous with respect to other samples or in regard to the context of the information
- **Irrelevant data**: data that does not affect the results of statistical or analitical methods
- **Inconsistent data**: data that can be represented in two or more ways
- **Malicious data**: data designed to cause damage or an undesired effect
- **Outliers**: data that is deviated from or inconsistent with other data in the dataset, also referred to as noise or anomolies

These three concepts will be elaborated on and used in the following investigation while preparing and preprocessing the data. 

In [1]:
# Imports
import string
import numpy as np
import pandas as pd
from IPython.core.display import display, HTML

# Increase the width of the notebook for displaying DataFrames
display(HTML("<style>.container { width:75% !important; }</style>"))

Known dependencies:

`pandas-0.23.4`

In [2]:
# Load the job market dataset into a pandas DataFrame
df = pd.read_csv("data.csv", dtype={"Id": np.str, "Location": np.str, "Area": np.str, 
                                    "Classification": np.str, "SubClassification": np.str})

## 1) Describing the Dataset

SEEK is a company that facilitates a platform for jobseekers to find jobs, and for employers to find employees. More specifically, they host one of Australia's largest employment marketplaces online under the url `www.seek.com.au`.

The provided dataset, which we will be using within this investigation, is of the SEEK job marketplace. Where each row is a job offering, identified by an Id, with information about the details, specifications and requirements of the job. Besides the job Id and Salary information, all data within the dataset is textual. Salary information is split into two columns, *LowestSalary* and *HighestSalary*, which are the only informative numerical columns in the dataset. 

In [3]:
df.head()

Id                                              Title  \
0  37404348                           Casual Stock Replenisher   
1  37404337                           Casual Stock Replenisher   
2  37404356  RETAIL SALES SUPERSTARS and STYLISTS Wanted - ...   
3  37404330                              Team member - Belrose   
4  37404308  Business Banking Contact Centre Specialist, Ni...   

                                          Company                      Date  \
0                                     Aldi Stores  2018-10-07T00:00:00.000Z   
1                                     Aldi Stores  2018-10-07T00:00:00.000Z   
2                             LB Creative Pty Ltd  2018-10-07T00:00:00.000Z   
3                          Anaconda Group Pty Ltd  2018-10-07T00:00:00.000Z   
4  Commonwealth Bank - Business & Private Banking  2018-10-07T00:00:00.000Z   

                  Location                         Area  \
0                   Sydney  North West & Hills District   
1    Richmond & Hawkesbury                          NaN   
2                 Brisbane          CBD & Inner Suburbs   
3  Gosford & Central Coast                          NaN   
4                   Sydney        Ryde & Macquarie Park   

                   Classification  SubClassification  \
0      Retail & Consumer Products  Retail Assistants   
1      Retail & Consumer Products  Retail Assistants   
2      Retail & Consumer Products  Retail Assistants   
3      Retail & Consumer Products  Retail Assistants   
4  Call Centre & Customer Service    Sales - Inbound   

                                         Requirement FullDescription  \
0  Our Casual Stock Replenishers pride themselves...             NaN   
1  Our Casual Stock Replenishers pride themselves...             NaN   
2  BRAND NEW FLAGSHIP STORE OPENING - SUNSHINE PLAZA             NaN   
3  Bring it on - do you love the great outdoors a...             NaN   
4  We are seeking highly articulate, enthusiastic...             NaN   

   LowestSalary  HighestSalary JobType  
0             0             30     NaN  
1             0             30     NaN  
2             0             30     NaN  
3             0             30     NaN  
4             0             30     NaN

### What are the categories/domains of the dataset?

This question is answered under the assumption that the categories/domains/variations of a dataset refer to the broad field in which the data is from. For example, the domain of a dataset of hospital patients would be healthcare, and the domain of a dataset containing information about stars and galaxies would be astronomy. 

Under our assumption, the domain of the SEEK dataset is employment marketplace. Where job offers are posted by companies and organisations for jobseekers to find.  

### What is the dataset size of each variation?

Based on the assumption stated above, our dataset consists of only one variation (namely the employment marketplace). Therefore, the dataset size of the employment marketplace variation can be found via the following cell:

In [4]:
# Print the dataset size of each variation
print("Size of SEEK dataset (employment marketplace variation): {} job listings".format(len(df)))

Size of SEEK dataset (employment marketplace variation): 318477 job listings


### What is the dataset structure/format? 

The dataset structure is that of a job description containing all of its listing information. More specifically, each row in the dataset is a job listing described by the following attributes: job title, company name, date, location, area within the location, general classification of the job and its specific sub-classification, requirements, full description, and the lowest and highest salary for the respective sub-classification.

The job title, location, area, classification and sub-classification are short and consistent strings that are part of a relatively small set of values. These values are likely taken from the SEEK websites selection of values. The company name and date are short strings as well.

Requirements and the full description of the job listing are longer strings, where the full description contains the raw html of the job description. They contain whitespace characters, unicode emojis and bulletpoint values. They are mostly inconsistent and are mostly all unique for each row.

The lowest salary and highest salary columns consist of 11 different integer values for each salary increment. The job type column contains 4 string values categorising each job listing as either part time, full time, contract/temp, or casual/vacation.

In [5]:
# Columns of the raw dataset
print("{:<20} {}".format("Column", "Datatype"))
print("{:<20} {}".format("------", "--------"))
for col in np.c_[list(df.columns), list(df.dtypes)]:
    print("{:<20} {}".format(col[0], col[1]))

Column               Datatype
------               --------
Id                   object
Title                object
Company              object
Date                 object
Location             object
Area                 object
Classification       object
SubClassification    object
Requirement          object
FullDescription      object
LowestSalary         int64
HighestSalary        int64
JobType              object


### Which parts of the dataset will be used? 

We hypothesise that we will be using all of the columns within the dataset for analysis and visulisation except for the id column which is irrelevant for our investigation purposes, as it does not offer anything significant in regard to job listings. However, since it uniquely identifies rows it will be kept for when operating with PySpark.

### Conclusions in regard to describing the dataset

The dataset to be used for this investigation is from the company SEEK's website. The website provides a platform for jobseeksers to find employers and employers to find their employees. Consequently, the website consists of a collection of job advestisement listings containing information about particular jobs. A crawler was used to collect this information and form the dataset to be used. 

In more detail, the dataset contains 318,477 job listings, each of which is described by the following attributes: id, title, company, date, location, area, classification, sub-classification, requirement, full description, lowest salary, highest salary, and job type. Due to the fact that this information is still raw within the dataset, since it assumedly has not been processed after being collected by the crawler, the data is dirty. Accordingly, the following section will describe the steps involved in preprocessing the data and preparing it for use in the analysis. 

## 2) Steps used for Data Preparation and Preprocessing

### Loading the dataset with Pandas

We use the Pandas function `pandas.read_csv()` to load the data into the notebook. The raw dataset contained inconsistent representations of the columns "Id", "Location", "Area", "Classification" and "SubClassification". In the case of "Id", the regular Id values were 8 numbers long and therefore interpreted as integers. However, there we 43462 job advertisement samples with Id values followed by a string of characters. For example:

```
       Valid ID: 37915260
Inconsistent ID: 37915260&searchrequesttoken=e859cc74-e22f-498d-ac7c-77a7e1b45676   
```

Consequently, Pandas was not able to read the the "Id" column values as integers and therefore stored them as objects (str). 

We outline the specific function and parameters used to read the dataset below:
```
df = pd.read_csv("data.csv", dtype={"Id": np.str, "Location": np.str, "Area": np.str, "Classification": np.str, "SubClassification": np.str})
```

For the other columns that raised the DtypeWarning it is assumed that their rows contained nan values of type float. As we found such occurrences while preprocessing the string data. 

### Data normalisation

As mentioned within *Introduction*, data normalization in the context of this course entails both normalizing the data distribution and scaling the data into a specific range. These normalization techniques are for numerical data. We therefore consider the numerical columns of the dataset, *Lowest Salary* and *Highest Salary*, and investigate the importance of normailzing these distributions. 

In [6]:
# Extract series from the DataFrame
lowest_salary_series = np.array(df["LowestSalary"].values)
highest_salary_series = np.array(df["HighestSalary"].values)

# Find min and max values of the two series
lowest_salary_min = min(lowest_salary_series)
lowest_salary_max = max(lowest_salary_series)
highest_salary_min = min(highest_salary_series)
highest_salary_max = max(highest_salary_series)

print(" LowestSalary range:  {} - {}".format(lowest_salary_min, lowest_salary_max))
print("HighestSalary range: {} - {}".format(highest_salary_min, highest_salary_max))

 LowestSalary range:  0 - 200
HighestSalary range: 30 - 999


In [7]:
# Scale Lowest Salary values into the range of 0.0 - 1.0
lowest_salary_scaled_series = np.zeros(lowest_salary_series.shape)
for i, val in enumerate(lowest_salary_series):
    lowest_salary_scaled_series[i] = (val - lowest_salary_min) / (highest_salary_max - lowest_salary_min)
    
# Scale Highest Salary values into the range of 0.0 - 1.0
highest_salary_scaled_series = np.zeros(highest_salary_series.shape)
for i, val in enumerate(highest_salary_series):
    highest_salary_scaled_series[i] = (val - lowest_salary_min) / (highest_salary_max - lowest_salary_min)

# Add the two scaled series to the DataFrame
df = df.assign(LowestSalaryScaled=lowest_salary_scaled_series)
df = df.assign(HighestSalaryScaled=highest_salary_scaled_series)

# Extract series from the DataFrame
lowest_salary_scaled_series = np.array(df["LowestSalaryScaled"].values)
highest_salary_scaled_series = np.array(df["HighestSalaryScaled"].values)

# Find min and max values of the two series
lowest_salary_scaled_min = min(lowest_salary_scaled_series)
lowest_salary_scaled_max = max(lowest_salary_scaled_series)
highest_salary_scaled_min = min(highest_salary_scaled_series)
highest_salary_scaled_max = max(highest_salary_scaled_series)

print(" LowestSalaryScaled range: {:.2f} - {:.2f}".format(lowest_salary_scaled_min, lowest_salary_scaled_max))
print("HighestSalaryScaled range: {:.2f} - {:.2f}".format(highest_salary_scaled_min, highest_salary_scaled_max))

 LowestSalaryScaled range: 0.00 - 0.20
HighestSalaryScaled range: 0.03 - 1.00


Within the investigation above, we found that the ranges of numerical data for the two columns in consideration were 0-200 and 30-999 for *Lowest Salary* and *Highest Salary*, respectively. It is assumed that the magnitude of this scale is slightly too large for distance fuctions in statistical analysis techniques. However, transforming the numerical scale of salary will remove the interpretable nature of the data which will be needed for data visualization. Consequenty, we decided to add two new columns to the dataset for the scaled lowest salary values (*LowestSalaryScaled*) and the scaled highest salary values (*HighestSalaryScaled*).

### Data cleaning

Within the *Introduction*, we outlined that data cleaning entails correcting and/or removing dirty data from the dataset. In the data cleaning analysis that follows, we clean the Id values, format the date samples, set the data type of columns in the DataFrame, and clean all textual data. Further reasoning is provided within each section below.

#### Cleaning Id values

As mentioned within *Loading the dataset with Pandas*, the raw dataset values of the "Id" column had inconsistencies in their representation. Most Id values were 8 number long integers, however there were other job advertisements with valid Id values followed by a string of characters. 

We identified that all of the strings following the Id values started within an ampersand ('&'). This allowed us to select the string of characters after the valid Id number using regular expressions, and remove them with the Pandas fuction `pandas.Series.replace`.

In [8]:
df["Id"] = df["Id"].replace(to_replace=r'&.*', value='', regex=True)

####  Formatting Date values

The raw dataset values of the "Date" column were represented in a format that contained both date and time information of when the advertisements were uploaded to SEEK. The website does not display specific time information, only showing the date. As a result, all time values were `00:00:00.000`. This information was consequently thought to be redundant. 

We identified that the Date values were all formatted in the following way:

`2018-10-07T00:00:00.000Z`

Where the date and time values were separated by the character 'T'. To remove the time value, we used regular expressions to select the 'T' as well as all subsequent characters within the string, and then replaced them with an empty string using the Pandas function `pandas.Series.replace`. 

In [9]:
df["Date"] = df["Date"].replace(to_replace=r'T.*', value='', regex=True)

#### Setting the data types of columns

We changed the data type of the column "Date" to datetime64[ns]. This was required due to formatting and consistency errors in the raw dataset. Furthermore, the data type conversion was made to provide a consistent representation of the data and allow for the use of functions and methods that process these data types. 

In [10]:
# Data types of raw dataset
df.dtypes

Id                      object
Title                   object
Company                 object
Date                    object
Location                object
Area                    object
Classification          object
SubClassification       object
Requirement             object
FullDescription         object
LowestSalary             int64
HighestSalary            int64
JobType                 object
LowestSalaryScaled     float64
HighestSalaryScaled    float64
dtype: object

In [11]:
# Set column "Id" to type int64
df["Id"] = pd.to_numeric(df["Id"])

# Set column "Date" to type datetime64[ns]
df["Date"] = pd.to_datetime(df["Date"])

In [12]:
# Data types after data cleaning and conversions
df.dtypes

Id                              int64
Title                          object
Company                        object
Date                   datetime64[ns]
Location                       object
Area                           object
Classification                 object
SubClassification              object
Requirement                    object
FullDescription                object
LowestSalary                    int64
HighestSalary                   int64
JobType                        object
LowestSalaryScaled            float64
HighestSalaryScaled           float64
dtype: object

#### Cleaning textual data

Raw textual data contains whitespace characters, unicode for special characters and emojis, and naturally, punctuation. Punctuation is useful for understanding a language, when the context and structure behind the words needs to be extracted. Within this investigation, we will not be investigating the meaning of the sentences as a whole, rather the meaning behind specific words. Consequently, punctuation, whitespace characters and special characters will not provide anything towards our analysis. As a result, we decided to remove these characters from our dataset to increase the efficiency of processing textual data during our anaylsis. In addition to the removal of the aforementioned characters, we replaced NaN values with empty strings and set all characters to lower case for the same reason as to increase our processing efficiency. The only exception to this was that the "FullDescription" column was left as it is, without any cleaning besides removing NaN values. This decision was made to preserve the HTML syntax found within the text of this attribute. The HTML syntax was later used in the analysis for parsing and was consequently left as is. 

In [13]:
# Print the number of null counts per dataset column
print("{:<20} {}".format("Column", "null count"))
for col in df.columns:
    print("{:<20} {}".format(col, df[col].isnull().sum()))

Column               null count
Id                   0
Title                0
Company              12004
Date                 0
Location             121248
Area                 195819
Classification       121248
SubClassification    121248
Requirement          7
FullDescription      16175
LowestSalary         0
HighestSalary        0
JobType              16098
LowestSalaryScaled   0
HighestSalaryScaled  0


In [14]:
# For each column in the DataFrame
for col in df.columns:
    
    # For each column of data type string 
    # Note: the "FullDescription column is left unchanged so that the HTML can be parsed in the analysis section
    if df[col].dtype == "object" and col != "FullDescription":
        
        # Clean whitespace characters
        df[col] = df[col].replace(to_replace=['\n', '\t', '\r'], value='')
        
        # Clean HTML unicode
        df[col] = df[col].replace(to_replace=[r'&#\d\d\d\d'], value='', regex=True)
        
        # Clean punctuation, replace NaN values with an empty string, and set all characters to lower case
        for r, row in enumerate(df[col].values):
                
            row = str(row)
            row = row.lower()
            
            if row == "nan" or pd.isnull(row):
                row = ""
                
            row = row.translate(str.maketrans('', '', string.punctuation))
            
            # Set the whitespace between words to be a single space
            tokens = row.split()
            row = " ".join(tokens)
            
            df.at[r, col] = row
            
# Replace NaN values with an empty string for the FullDescription column
for r, row in enumerate(df["FullDescription"].values):
    if row == "nan" or pd.isnull(row):
        row = ""
        df.at[r, "FullDescription"] = row

# Replace emojis with empty strings in the titles
emoji_regex_string = r'(💝|📣|💥|👗|👕|🛍|👉|👟|🏆|👑|🎄|🎁|🎉|🎗|🌎|➕|✨|⭐️|⚡️|★|🏖|❗️)'
df["Title"] = df["Title"].replace(to_replace=[emoji_regex_string], value="", regex=True)

In [15]:
# Print the number of null counts per dataset column to check they have been cleaned
print("{:<20} {}".format("Column", "null count"))
for col in df.columns:
    print("{:<20} {}".format(col, df[col].isnull().sum()))

Column               null count
Id                   0
Title                0
Company              0
Date                 0
Location             0
Area                 0
Classification       0
SubClassification    0
Requirement          0
FullDescription      0
LowestSalary         0
HighestSalary        0
JobType              0
LowestSalaryScaled   0
HighestSalaryScaled  0


In [16]:
df.tail()

Id                                          Title  \
318472  38564891                program manager trade reporting   
318473  38564940          web content writer june 2019 contract   
318474  38552964            brand director global premium brand   
318475  38534438  head of financial planning reporting strategy   
318476  38561751                         head of operations eoi   

                                                  Company       Date Location  \
318472                                            talenza 2019-03-13            
318473  talent – winner ‘seek large recruitment agency... 2019-03-13            
318474                   retail career consulting pty ltd 2019-03-12            
318475                                         moir group 2019-03-08            
318476                                 austcorp executive 2019-03-13            

       Area Classification SubClassification  \
318472                                         
318473                                         
318474                                         
318475                                         
318476                                         

                                              Requirement FullDescription  \
318472  our client is a large financial services firm ...                   
318473  you will be able to write clearly and concisel...                   
318474  newly created brand director aus must be comme...                   
318475  highquality manufacturing organisation excitin...                   
318476      expressions of interest for a leadership role                   

        LowestSalary  HighestSalary JobType  LowestSalaryScaled  \
318472           200            999                      0.2002   
318473           200            999                      0.2002   
318474           200            999                      0.2002   
318475           200            999                      0.2002   
318476           200            999                      0.2002   

        HighestSalaryScaled  
318472                  1.0  
318473                  1.0  
318474                  1.0  
318475                  1.0  
318476                  1.0

In [17]:
# Save the cleaned and preprocessed dataset
df.to_csv("preprocessed_data.csv")

### Conclusions in regard to data preparation and preprocessing

Within this section we applied data preparation and preprocessing techniques to the SEEK dataset in order to prepare the data for analysis and interpretation. The raw comma-separated values (CSV) file used for storing the dataset contained representation inconsistencies, where numerical data contained charater data, and character data assumedly contained numerical NaN values. Loading this dirty data therefore required telling pandas to load all the attributes as strings. These inconsistent representations were then cleaned in the subsequent preprocesssing steps. 

The lowest salary and highest salary attributes were the only attributes that contained meaningful numerical data in the dataset. We therefore applied data normalisation techniques to these attributes by scaling the existing range of 0 to 999 into the range of 0.00 to 1.00 such that statistical analysis techniques could be accurately used on the salary ranges.

To correct and/or remove dirty data from the dataset, and to best prepare the data for data analysis and interpretation, data cleaning techniques were used. Firstly, redundant date representations were found in the dataset and cleaned through the use of regular expressions. Secondly, the id column was cleaned and converted into type int64, and the data type of the date column was converted from object (str) into datetime64[ns] such that functions and methods requiring this data type format could be used within the analysis. Lastly, all of the textual data within the dataset was cleaned and prepared for analysis by cleaning whitespace, punctuation and unicode characters, by replacing NaN values with empty strings, and by setting all characters to lower case. Additionally, specific emojies were replaced with empty strings in the text of the title attribute. Throughout this textual data cleaning process, the non-NaN text within the full description attribute was left unchanged. This decision was made to preserve the raw HTML contents of the text, such that it could be later utilised in the analysis for parsing specific information.

These preparation and preprocessing techniques adequately prepared the data within the dataset for analysis and interpretation. Thus, the preprocessed dataset was saved to file so that it could be used within the analysis notebook. 

## 3) Hypotheses regarding the Analysis

We firstly consider an analysis on the location attribute of the dataset. When investigating the job market share across the cities of Australia, it is hypothesised that the majority of the market share will be in the major cities of Australia, with more rural and suburban areas having significantly less jobs advertised on the employment marketplace. Furthermore, we hypothesise that cities will have higher average salaries than other rural and suburban areas in Australia, since cities have competitive job markets that require employers to increase salaries in order to secure employees over competitors.

In relation to the trends of the job market, we expect that the Information Communication Technology (ICT) sector has the largest market share among all sectors within the database. Moreover, we hypothesise that the average salary of the ICT sector has increased over the duration of the dataset, which is from October 2018 to March 2019, due to recent technological advancements and in turn investments in the ICT sector by companies and governmental organisations. When investigating the trends of posting job advertisements, we expect that the majority of jobs will be posted throughout the week, as employees generally are not working on weekends and therefore would not be posting job listings for their company.

Regarding the salaries of job advertisements within sub-sectors, we hypothesise that jobs within sub-sectors requiring a higher level of expertise (such as Criminal Civil Law, Oil Gas Drilling or Automotive Engineering) will have higher salary ranges than sub-sectors with minimal entry requirements (such as Floristry or Nannies/Babysitters). Furthermore, we believe that some sectors such as ICT and Engineering will have a diverse range of job listing salaries. Since there are jobs within these sectors that could be low paying, or that could be in the top salary range among all job listings. In contrast, for sectors such as Call Centre Customer Service and Retail Consumer Products, it is assumed that the majority of job offerings within these sectors will be in the lower salary ranges due to the fact that the entry level of expertise for these positions is not very high. 

Our final hypothesis is in relation to the textual analysis to be conducted. Our aim is to extract skillsets from the job descriptions so that we can determine what are the most frequently requested skills for applicants within each sector. Since the requirement attribute of the provided dataset is not accurate in regard to skillsets (i.e. the textual data under the attribute contains other information than just job requirements), we believe that we will have to use the full description attribute to construct our own textual dataset regarding skillsets and requirements. Since this data is by nature dirty, and because we are not using any advanced natural language processing techniques due to the time constraints of the assessment, we hypothesise that the skillsets we find for each sector may not be perfectly accurate in regard to the context of skills required for a given sector. However, we still believe that we will be able to provide an informative study on the skillsets that are popular among different sectors and within the job market as a whole. 